Add bulk guide (#155)
opensearch-project/opensearch-rs: guides/bulk.md

### Creating multiple documents

The following code creates two documents in the `movies` index and one document in the `books` index:

```rust
client
    .bulk(BulkParts::None)
    .body(Vec::<JsonBody<_>>::from([
        json!({ "create": { "_index": movies, "_id": "1" }}).into(),
        json!({ "title": "Beauty and the Beast", "year": "1991" }).into(),
        json!({ "create": { "_index": movies, "_id": "2" }}).into(),
        json!({ "title": "Beauty and the Beast - Live Action", "year": "2017" }).into(),
        json!({ "create": { "_index": books, "_id": "1" }}).into(),
        json!({ "title": "The Lion King", "year": "1994" }).into(),
    ]))
    .send()
    .await?;
```

As you can see, each bulk operation is comprised of two objects. The first object contains the operation type and the target document's `_index` and `_id`. The second object contains the document's data. As a result, the body of the request above contains six objects for three index actions.

> Alternatively, the `bulk` method can accept `BulkOperations` where each represents a single `BulkOperation`. The following code is equivalent to the previous example:

Reviewer: This phrasing does not make sense, a more relevant descriptor is:
```suggestion
Alternatively, you may use the `BulkOperation` helper type along with its `BulkOperations` container type as a type-safe abstraction over this format. The following code is equivalent to the previous example:
```

# Bulk

In this guide, you'll learn how to use the OpenSearch Rust Client API to perform bulk operations. You'll learn how to index, update, and delete multiple documents in a single request.

## Setup

First, create a client instance with the following code:

```rust
let url = Url::parse("https://localhost:9200")?;
let credentials = Credentials::Basic("admin".into(), "admin".into());
let transport = TransportBuilder::new(SingleNodeConnectionPool::new(url))
    .cert_validation(CertificateValidation::None)
    .auth(credentials)
    .build()?;
let client = OpenSearch::new(transport);
```

Next, create an index named `movies` and another named `books` with the default settings:

```rust
let movies = "movies";
let books = "books";
client.indices().create(IndicesCreateParts::Index(movies)).send().await?;
client.indices().create(IndicesCreateParts::Index(books)).send().await?;
```

## Bulk API

The `bulk` API action allows you to perform document operations in a single request. The body of the request is an array of objects that contains the bulk operations and the target documents to index, create, update, or delete.

> ### Indexing multiple documents

Reviewer: Due to the semantics of "create" generally assuming an id is being specified, I believe it makes sense to swap the operation types of this section with the next one.
```suggestion
### Creating multiple documents
```

The following code creates two documents in the `movies` index and one document in the `books` index:

```rust
client
    .bulk(BulkParts::None)
    .body(Vec::<JsonBody<_>>::from([

>         json!({ "index": { "_index": movies, "_id": "1" }}).into(),
>         json!({ "title": "Beauty and the Beast", "year": "1991" }).into(),
>         json!({ "index": { "_index": movies, "_id": "2" }}).into(),
>         json!({ "title": "Beauty and the Beast - Live Action", "year": "2017" }).into(),
>         json!({ "index": { "_index": books, "_id": "1" }}).into(),
>         json!({ "title": "The Lion King", "year": "1994" }).into(),

Reviewer: Let's make these create operations rather than index operations
```suggestion
        json!({ "create": { "_index": movies, "_id": "1" }}).into(),
        json!({ "title": "Beauty and the Beast", "year": "1991" }).into(),
        json!({ "create": { "_index": movies, "_id": "2" }}).into(),
        json!({ "title": "Beauty and the Beast - Live Action", "year": "2017" }).into(),
        json!({ "create": { "_index": books, "_id": "1" }}).into(),
        json!({ "title": "The Lion King", "year": "1994" }).into(),
```

    ]))
    .send()
    .await?;
```

As you can see, each bulk operation is comprised of two objects. The first object contains the operation type and the target document's `_index` and `_id`. The second object contains the document's data. As a result, the body of the request above contains six objects for three index actions.

Alternatively, the `bulk` method can accept an array of hashes where each hash represents a single operation. The following code is equivalent to the previous example:

```rust

> let mut ops = BulkOperations::new();
> ops.push(
>     BulkOperation::index(json!({
>         "title": "Beauty and the Beast",
>         "year": "1991"
>     }))
>     .index(movies)
>     .id("1"),
> )?;
> ops.push(
>     BulkOperation::index(json!({
>         "title": "Beauty and the Beast - Live Action",
>         "year": "2017"
>     }))
>     .index(movies)
>     .id("2"),
> )?;
> ops.push(
>     BulkOperation::index(json!({
>         "title": "The Lion King",
>         "year": "1994"
>     }))
>     .index(books)
>     .id("1"),
> )?;
> 
> client.bulk(BulkParts::None).body(vec![ops]).send().await?;

Reviewer: Let's make these create operations to match above.
```suggestion
let mut ops = BulkOperations::new();
ops.push(
    BulkOperation::create("1", json!({
        "title": "Beauty and the Beast",
        "year": "1991"
    }))
    .index(movies),
)?;
ops.push(
    BulkOperation::create("2",  json!({
        "title": "Beauty and the Beast - Live Action",
        "year": "2017"
    }))
    .index(movies),
)?;
ops.push(
    BulkOperation::create("1", json!({
        "title": "The Lion King",
        "year": "1994"
    }))
    .index(books),
)?;

client.bulk(BulkParts::None).body(vec![ops]).send().await?;
```

```

We will use this format for the rest of the examples in this guide.

> ### Creating multiple documents
> 
> Similarly, instead of calling the `create` method for each document, you can use the `bulk` API to create multiple documents in a single request. The following code creates three documents in the `movies` index and one in the `books` index:

Reviewer: Let's swap this to the `index` operation.

```suggestion
### Indexing multiple documents

Similarly, instead of calling the `index` method for each document, you can use the `bulk` API to index multiple documents in a single request. The following code indexes three documents in the `movies` index and one in the `books` index:
```

```rust

> let mut ops = BulkOperations::new();
> ops.push(
>     BulkOperation::create(json!({
>         "title": "Beauty and the Beast 2",
>         "year": "2030"
>     }))
>     .index(movies),
> )?;
> ops.push(
>     BulkOperation::create(json!({
>         "title": "Beauty and the Beast 3",
>         "year": "2031"
>     }))
>     .index(movies),
> )?;
> ops.push(
>     BulkOperation::create(json!({
>         "title": "Beauty and the Beast 4",
>         "year": "2049"
>     }))
>     .index(movies),
> )?;
> ops.push(
>     BulkOperation::create(json!({
>         "title": "The Lion King 2",
>         "year": "1998"
>     }))
>     .index(books),
> )?;
> client.bulk(BulkParts::None).body(vec![ops]).send().await?;

Reviewer: Let's swap these to `index` operation, and also `movies` should only specified on the bulk call, and only specifying `books` specially on the one operation. 
```suggestion
let mut ops = BulkOperations::new();
ops.push(
    BulkOperation::index(json!({
        "title": "Beauty and the Beast 2",
        "year": "2030"
    })),
)?;
ops.push(
    BulkOperation::index(json!({
        "title": "Beauty and the Beast 3",
        "year": "2031"
    })),
)?;
ops.push(
    BulkOperation::index(json!({
        "title": "Beauty and the Beast 4",
        "year": "2049"
    })),
)?;
ops.push(
    BulkOperation::index(json!({
        "title": "The Lion King 2",
        "year": "1998"
    }))
    .index(books),
)?;

client.bulk(BulkParts::Index(movies)).body(vec![ops]).send().await?;
```

```

> Note that we specified only the `_index` for the last document in the request body. This is because the `bulk` method accepts an `index` parameter that specifies the default `_index` for all bulk operations in the request body. Moreover, we omit the `_id` for each document and let OpenSearch generate them for us in this example, just like we can with the `create` method.

Reviewer: ```suggestion
Note that we specified only the `_index` for the last document in the request body. This is because the `bulk` method accepts an `index` parameter that specifies the default `_index` for all bulk operations in the request body. Moreover, we omit the `_id` for each document and let OpenSearch generate them for us in this example, just like we can with the `index` method.
```

### Updating multiple documents

```rust

> let mut ops = BulkOperations::new();
> ops.push(
>     BulkOperation::update(json!({
>         "year": "1992"
>     }))
>     .index(movies)
>     .id("1"),
> )?;
> ops.push(
>     BulkOperation::update(json!({
>         "year": "2018"
>     }))
>     .index(movies)
>     .id("2"),
> )?;
> 
> client.bulk(BulkParts::None).body(vec![ops]).send().await?;

Reviewer: The id gets specified in the `update` call, additionally the `doc` nesting is still required, and the index should just be specified in the bulk call.
```suggestion
let mut ops = BulkOperations::new();
ops.push(BulkOperation::update(
    "1",
    json!({
        "doc": {
            "year": "1992"
        }
    }),
))?;
ops.push(BulkOperation::update(
    "2",
    json!({
        "doc": {
            "year": "2018"
        }
    }),
))?;

client.bulk(BulkParts::Index(movies)).body(vec![ops]).send().await?;
```

```

> 

Reviewer: The note needs re-adding about the `doc` field
```suggestion

Note that the updated data is specified in the `doc` field of the source object.

```

### Deleting multiple documents

```rust

> let mut ops = BulkOperations::new();
> ops.push(
>     BulkOperation::delete()
>     .index(movies)
>     .id("1"),
> )?;
> ops.push(
>     BulkOperation::delete()
>     .index(movies)
>     .id("2"),
> )?;
> 
> client.bulk(BulkParts::None).body(vec![ops]).send().await?;

Reviewer: The id gets specified in the `delete` call, and index can be moved to `bulk` call
```suggestion
let mut ops = BulkOperations::new();
ops.push(BulkOperation::<()>::delete("1"))?;
ops.push(BulkOperation::<()>::delete("2"))?;

client.bulk(BulkParts::Index(movies)).body(vec![ops]).send().await?;
```

```

### Mix and match operations

You can mix and match the different operations in a single request. The following code creates two documents, updates one document, and deletes another document:

```rust

> let mut ops = BulkOperations::new();
> 
> ops.push(
>     BulkOperation::create(json!({
>         "title": "Beauty and the Beast 5"
>         "year": "2050"
>     }))
>     .index(movies),
> )?;
> ops.push(
>     BulkOperation::create(json!({
>         "title": "Beauty and the Beast 6"
>         "year": "2051"
>     }))
>     .index(movies),
> )?;
> ops.push(
>     BulkOperation::update(json!({
>         "year": "2052"
>     }))
>     .index(movies)
>     .id("3"),
> )?;
> ops.push(
>     BulkOperation::delete()
>     .index(movies)
>     .id("4"),
> )?;
> 
> client.bulk(BulkParts::None).body(vec![ops]).send().await?;

Reviewer: Bring this inline with previous comments about using each operation:
```suggestion
let mut ops = BulkOperations::new();

ops.push(BulkOperation::index(json!({
    "title": "Beauty and the Beast 5",
    "year": "2050"
})))?;
ops.push(BulkOperation::index(json!({
    "title": "Beauty and the Beast 6",
    "year": "2051"
})))?;
ops.push(BulkOperation::update(
    "3",
    json!({
        "doc": {
            "year": "2052"
        }
    }),
))?;
ops.push(BulkOperation::<()>::delete("4"))?;

client.bulk(BulkParts::Index(movies)).body(vec![ops]).send().await?;
```

```

### Handling errors

The `bulk` API returns an array of responses for each operation in the request body. Each response contains a `status` field that indicates whether the operation was successful or not. If the operation was successful, the `status` field is set to a `2xx` code. Otherwise, the response contains an error message in the `error` field.

The following code shows how to look for errors in the response:

```rust

> let mut ops = BulkOperations::new();
> 
> ops.push(
>     BulkOperation::create(json!({
>         "title": "Beauty and the Beast"
>         "year": "1991"
>     }))
>     .index(movies)
>     .id("1"),
> )?;
> ops.push(
>     BulkOperation::create(json!({
>         "title": "Beauty and the Beast 2"
>         "year": "2030"
>     }))
>     .index(movies)
>     .id("2"),
> )?;
> ops.push(
>     BulkOperation::create(json!({
>         "title": "Beauty and the Beast 3"
>         "year": "2031"
>     }))
>     .index(movies)
>     .id("1"),
> )?;
> ops.push(
>     BulkOperation::create(json!({
>         "title": "Beauty and the Beast 4"
>         "year": "2049"
>     }))
>     .index(movies)
>     .id("2"),
> )?;
> 
> let response = client
>   .bulk(BulkParts::None)
>   .body(vec![ops])
>   .send()
>   .await?;
> 
> let response_body = response.json::<Value>().await?;
> 
> for item in response["items"].iter() {
>   let range  = 200..299;
>   if !(200..299).contains(&item["create"]["status"].as_u64().unwrap()) {
>     println!("{}", item["create"]["error"]["reason"]);
>   }
> }

Reviewer: Same fixes again

```suggestion
let mut ops = BulkOperations::new();

ops.push(BulkOperation::create(
    "1",
    json!({
        "title": "Beauty and the Beast",
        "year": "1991"
    }),
))?;
ops.push(BulkOperation::create(
    "2",
    json!({
        "title": "Beauty and the Beast 2",
        "year": "2030"
    }),
))?;
ops.push(BulkOperation::create(
    "1",
    json!({
        "title": "Beauty and the Beast 3",
        "year": "2031"
    }),
))?;
ops.push(BulkOperation::create(
    "2",
    json!({
        "title": "Beauty and the Beast 4",
        "year": "2049"
    }),
))?;

let response = client
    .bulk(BulkParts::Index(movies))
    .body(vec![ops])
    .send()
    .await?
    .json::<Value>()
    .await?;

for item in response["items"].as_array().unwrap() {
    if !(200..299).contains(&item["create"]["status"].as_u64().unwrap()) {
        println!("{}", item["create"]["error"]["reason"]);
    }
}
```

## Setup

First, create a client instance with the following code:

```rust
let url = Url::parse("https://localhost:9200")?;
let credentials = Credentials::Basic("admin".into(), "admin".into());
let transport = TransportBuilder::new(SingleNodeConnectionPool::new(url))
    .cert_validation(CertificateValidation::None)
    .auth(credentials)
    .build()?;
let client = OpenSearch::new(transport);
```

Next, create an index named `movies` and another named `books` with the default settings:

```rust
let movies = "movies";
let books = "books";
client.indices().create(IndicesCreateParts::Index(movies)).send().await?;
client.indices().create(IndicesCreateParts::Index(books)).send().await?;
```

## Bulk API

The `bulk` API action allows you to perform document operations in a single request. The body of the request is an array of objects that contains the bulk operations and the target documents to index, create, update, or delete.

### Indexing multiple documents

The following code creates two documents in the `movies` index and one document in the `books` index:

```rust
client
    .bulk(BulkParts::None)
    .body(Vec::<JsonBody<_>>::from([

>         json!({ "index": { "_index": movies, "_id": "1" }}).into(),

Reviewer 2: The index should be in quotes as well, right?

```suggestion
        json!({ "index": { "_index": "movies", "_id": "1" }}).into(),
```

Reviewer 2: Oh sorry, my bad, I see we use the variables from above, please nvd

        "title": "Beauty and the Beast 3",
        "year": "2031"
    }))
    .index(movies),
)?;
ops.push(
    BulkOperation::create(json!({
        "title": "Beauty and the Beast 4",
        "year": "2049"
    }))
    .index(movies),
)?;
ops.push(
    BulkOperation::create(json!({
        "title": "The Lion King 2",
        "year": "1998"
    }))
    .index(books),
)?;
client.bulk(BulkParts::None).body(vec![ops]).send().await?;
```

Note that we specified only the `_index` for the last document in the request body. This is because the `bulk` method accepts an `index` parameter that specifies the default `_index` for all bulk operations in the request body. Moreover, we omit the `_id` for each document and let OpenSearch generate them for us in this example, just like we can with the `create` method.

### Updating multiple documents

```rust
let mut ops = BulkOperations::new();
ops.push(
    BulkOperation::update(json!({
        "year": "1992"
    }))
    .index(movies)
    .id("1"),
)?;
ops.push(
    BulkOperation::update(json!({
        "year": "2018"
    }))
    .index(movies)
    .id("2"),
)?;

client.bulk(BulkParts::None).body(vec![ops]).send().await?;
```

> Note that the updated data is specified in the `doc` field of the `data` object.

Reviewer 2: I think this is not relevant here (since `BulkOperation::update` hides this fact), may be you could remove it or use `Json!` to demonstrate that?
